pkg/router/template/util/rewritetarget/rewritetarget_test.go

```diff
@@ -85,7 +85,35 @@ func Test_SanitizeInput(t *testing.T) {
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			got := rewritetarget.SanitizeInput(tc.input)
+			got := rewritetarget.SanitizeRewriteTargetInput(tc.input)
+			if got != tc.output {
+				t.Errorf("Failure: expected %s, got %s", tc.output, got)
+			}
+		})
+	}
+}
+
+func Test_SanitizeRewritePathInput(t *testing.T) {
+	testCases := []struct {
+		name   string
+		input  string
+		output string
+	}{
+		{
+			name:   "single quotes should be always be escaped",
+			input:  `'foo'foo\'`,
+			output: `'\''foo'\''foo\'\''`,
```

Comment:
`\'` ends up generating an unpaired quote. Doesn't that still break the template?

Reply:
So `\'` generates `\'\''` because our rule is simple & universal: `'` gets translated to `'\''`, and we don't do anything with the `\`, so you mostly disregard it.

To explain the `'\''` sequence:
1. The first `'` terminates the outer quote provided by the template. 
2. The `\'` specifies a literal escaped quote (it must be outside of the protection of `''`)
3. The last `'` starts the outer quote again.

So the answer is no: it doesn't break the template, the `\` in front of the quote doesn't matter. The result of `\'` is that it gets interpreted as literal `\'`, which is desirable in this specific situation.

I'll also mention that we did handle single quotes in a different way in the fix for https://issues.redhat.com/browse/OCPBUGS-22739, but the key different here is that, in OCPBUGS-22739, `\'` was a valid rewrite annotation sequence (it matched `'`), where for this fix, though it doesn't cause a syntax error, it causes an unreachable situation, because the HAProxy map files MUST match the same pattern as the replace-path regex. The [GenerateRouteRegexp](https://github.com/openshift/router/blob/fb70ac4d547954e7f96c12b4bc329917f1695d35/pkg/router/template/util/util.go#L53) function is the reason why the HAPRoxy map path regexs are interpreted literally. I updated https://docs.google.com/document/d/1wA6ZdRYaiN70TY5UbTOn1wl5qAn_NPb2ty_G85XhnYM/edit with more info on this situation where the map files don't match the `<match-regex>` argument for rewrite target.

I'm on a bit of a tangent in this comment. I also realized `GenerateRouteRegexp` used `regexp.QuoteMeta(path)` instead of `\Q...\E`. They appear to do the same exact thing, but I think it's better to be consistent, especially if there is a subtle different between the two. We need to have the path interpreted in the same exact way for rewrites to work.

So I'll update the code to use `QuoteMeta`, and I also found that you can add a newline character in spec.path and break this config. So I fixed that too.